File: [BOJ] 백준 풀이/단계별로 풀어보기 (step by step)/#9 재귀/2447_별 찍기 (22.04.10 성공).py
def solution(arr):
    tmparr = []
    length = len(arr) # N 크기

    # 새로이 그려야 하는데, 행과 열이 3개씩 더 있으므로 3을 곱한 만큼 반복
    for i in range(3 * length):
        if i // length == 1: # 중앙에 공백을 채우기 위한 장치 (N = 9일 때 3, 4, 5 공백)
            tmparr.append(arr[i % length] + " " * length + arr[i % length])
        else: # 이외에는 그냥 출력
            tmparr.append(arr[i % length] * 3)

    return tmparr
 
arr = ["***","* *","***"]
N = int(input())

# 그리기
while N != 3:
    arr = solution(arr)
    N = N // 3

# 결과 출력
for i in arr:
    print(i)

File: [BOJ] 백준 풀이/단계별로 풀어보기 (step by step)/#9 재귀/2447_별 찍기 (22.04.10 성공).cpp
#include <iostream>

using namespace std;

void solution(int i, int j, int N) {
    if ((i / N) % 3 == 1 && (j / N) % 3 == 1)  // 이 조건일 때 빈칸
        cout << " ";
    // 이미 위 조건에서 걸러진다는 것은 별이 찍혀야 한다는 의미인데, 이 문제를
    // 재귀적으로 해결하기 위해서는 재귀 조건이 있어야 한다. 현재 정사각형의
    // 크기가 9 이상이면 (N이 3 이상이면) 탐색할 수 있는 공백이 있지만 1이라면 공백이 없으므로
    // 그대로 *을 출력한다. (N / 3 == 0과 같다)
    else if (N == 1)
        cout << "*";
    else
        solution(i, j, N / 3);
}

int main() {
    int N;
    cin >> N;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) 
            solution(i, j, N);

        cout << "\n";
    }
}